compat/visual-composer/visual-composer.php

```diff
@@ -0,0 +1,282 @@
+		if ( empty( $parsed_value ) ) {
+			//Get the first value as the default.
+			reset( $so_widget_names );
+			$widget_class = key( $so_widget_names );
+		} else {
+			$widget_class = $parsed_value['widget_class'];
+		}
+
+		$widget = ! empty( $wp_widget_factory->widgets[ $widget_class ] ) ? $wp_widget_factory->widgets[ $widget_class ] : false;
+
+		ob_start();
+		$select->render( $widget_class ); ?>
+		<input type="hidden" name="ajaxurl"
+		       data-ajax-url="<?php echo wp_nonce_url( admin_url( 'admin-ajax.php' ), 'sowb_vc_widget_render_form', '_sowbnonce' ) ?>">
+		<input type="hidden" name="so_widget_data" class="wpb_vc_param_value" value="<?php echo esc_attr( $value ); ?>">
+		<div class="siteorigin_widget_form_container">
+			<?php
+			if ( ! empty( $widget ) && is_object( $widget ) && is_subclass_of( $widget, 'SiteOrigin_Widget' ) ) {
+				/* @var $widget SiteOrigin_Widget */
+				$widget->form( $parsed_value['widget_data'] );
+			}
+			?>
+		</div>
+		<?php
+		return ob_get_clean();
+	}
+
+	function sowb_vc_widget_render_form() {
+		if ( empty( $_REQUEST['widget'] ) ) {
+			wp_die();
+		}
+		if ( empty( $_REQUEST['_sowbnonce'] ) || ! wp_verify_nonce( $_REQUEST['_sowbnonce'], 'sowb_vc_widget_render_form' ) ) {
+			wp_die();
+		}
+
+		$request      = array_map( 'stripslashes_deep', $_REQUEST );
+		$widget_class = $request['widget'];
+
+		global $wp_widget_factory;
+
+		$widget = ! empty( $wp_widget_factory->widgets[ $widget_class ] ) ? $wp_widget_factory->widgets[ $widget_class ] : false;
+
+		if ( ! empty( $widget ) && is_object( $widget ) && is_subclass_of( $widget, 'SiteOrigin_Widget' ) ) {
+			/* @var $widget SiteOrigin_Widget */
+			$widget->form( array() );
+		}
+
+		wp_die();
+	}
+
+	function update_widget_data( $content ) {
+
+		$content = preg_replace_callback( '/\[siteorigin_widget_vc [^\]]*\]/', array(
+			$this,
+			'update_shortcode'
+		), $content );
+
+		return $content;
+	}
+
+	function update_shortcode( $shortcode ) {
+
+		preg_match( '/so_widget_data="([^"]*)"/', stripslashes( $shortcode[0] ), $widget_json );
+
+		// We double encode in the front end to prevent accidental decoding when the content is set on the
+		// WP visual editor.
+		$widget_json = html_entity_decode( html_entity_decode( $widget_json[1] ) );
+
+		$widget_atts = json_decode( $widget_json, true );
+
+		global $wp_widget_factory;
+
+		$widget = ! empty( $wp_widget_factory->widgets[ $widget_atts['widget_class'] ] ) ? $wp_widget_factory->widgets[ $widget_atts['widget_class'] ] : false;
+
+		if ( ! empty( $widget ) && is_object( $widget ) && is_subclass_of( $widget, 'SiteOrigin_Widget' ) ) {
+			/* @var $widget SiteOrigin_Widget */
+			$widget_atts['widget_data'] = $widget->update( $widget_atts['widget_data'], $widget_atts['widget_data'] );
+		}
+
+		$widget_json = json_encode( $widget_atts );
+
+		$widget_json = htmlentities( htmlentities( $widget_json ) );
+
+		$widget_json = str_replace( array(
```

Comment:
Similar style comment to earlier. This would be much clearer with arguments on their own lines. Unless there's some sort of style guideline that says we should be doing it the current way.

```
$widget_json = str_replace(
    array( '[', ']', ),
    array( '&#91;', '&#93;',
    $widget_json
);
```